xds/internal/client/helper.go

```diff
@@ -68,95 +58,90 @@ type Config struct {
-		Creds:        grpc.WithCredentialsBundle(google.NewComputeEngineCredentials()),
-		NodeProto:    bsData.node,
-	}, nil
+type channelCreds struct {
+	Type   string          `json:"type"`
+	Config json.RawMessage `json:"config"`
 }
 
-// bootstrapData wraps the contents of the bootstrap file.
-// Today the bootstrap file contains a Node proto and an ApiConfigSource proto
-// in JSON format.
-type bootstrapData struct {
-	node      *basepb.Node
-	xdsServer *cspb.ApiConfigSource
+type xdsServer struct {
+	ServerURI    string         `json:"server_uri"`
+	ChannelCreds []channelCreds `json:"channel_creds,omitempty"`
 }
 
-func (bsd *bootstrapData) balancerName() string {
-	// If the bootstrap file was read and parsed successfully, this should be
-	// setup properly. So, we skip checking for the presence of these fields
-	// before accessing and returning it.
-	return bsd.xdsServer.GetGrpcServices()[0].GetGoogleGrpc().GetTargetUri()
-}
+// NewConfig returns a new instance of Config initialized by reading the
+// bootstrap file found at ${GRPC_XDS_BOOTSTRAP}.
+//
+// The format of the bootstrap file will be as follows:
+// {
+//    "xds_server": {
+//      "server_uri": <string containing URI of xds server>,
+//      "channel_creds": [
+//        {
+//          "type": <string containing channel cred type>,
+//          "config": <JSON object containing config for the type>
+//        }
+//      ]
+//    },
+//    "node": <JSON form of basepb.Node proto>
+// }
+//
+// Currently, we support exactly one type of credential, which is
+// "google_default", where we use the host's default certs for transport
+// credentials and a Google oauth token for call credentials.
+//
+// This function tries to process as much of the bootstrap file as possible (in
+// the presence of the errors) and may return a Config object with certain
+// fields left unspecified, in which case the caller should use some sane
+// defaults.
+func NewConfig() *Config {
+	config := &Config{}
+
+	fName, ok := os.LookupEnv(bootstrapFileEnv)
+	if !ok {
+		grpclog.Errorf("xds: %s environment variable not set", bootstrapFileEnv)
+		return config
+	}
 
-func readBootstrapFile(name string) (*bootstrapData, error) {
-	grpclog.Infof("xds: Reading bootstrap file from %s", name)
-	data, err := fileReadFunc(name)
+	grpclog.Infof("xds: Reading bootstrap file from %s", fName)
+	data, err := fileReadFunc(fName)
 	if err != nil {
-		return nil, fmt.Errorf("xds: bootstrap file {%v} read failed: %v", name, err)
+		grpclog.Errorf("xds: bootstrap file {%v} read failed: %v", fName, err)
+		return config
 	}
 
 	var jsonData map[string]json.RawMessage
 	if err := json.Unmarshal(data, &jsonData); err != nil {
-		return nil, fmt.Errorf("xds: json.Unmarshal(%v) failed during bootstrap: %v", string(data), err)
+		grpclog.Errorf("xds: json.Unmarshal(%v) failed during bootstrap: %v", string(data), err)
+		return config
 	}
 
-	bsd := &bootstrapData{}
 	m := jsonpb.Unmarshaler{}
```

Comment:
This needs to also set `AllowUnknownFields: true`. So when bootstrap file contains more fields because newer client supports them, an old client won't fail on parsing.

Reply:
Done.